interfaces/builtin/cups.go

```diff
@@ -48,15 +59,139 @@ const cupsConnectedPlugAppArmor = `
+	// path contains only $SNAP, $SNAP_DATA, $SNAP_COMMON, and no
+	// other $VARs are present. It is ok to use
+	// ExpandSnapVariables() since it only expands $SNAP, $SNAP_DATA
+	// and $SNAP_COMMON
+	return snapInfo.ExpandSnapVariables(cupsdSocketSourceDir), nil
+}
+
+func (iface *cupsInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// verify that the snap has a cups-socket-dir interface attribute, which is
+	// needed to identify where to find the cups socket is located in the snap
+	// providing the cups socket
+	_, err := validateCupsSocketDirAttr(slot, slot.Snap)
+	return err
+}
+
+func (iface *cupsInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	cupsdSocketSourceDir, err := validateCupsSocketDirAttr(slot, slot.Snap())
+	if err != nil {
+		return err
+	}
+
+	// add the base snippet
+	spec.AddSnippet(cupsConnectedPlugAppArmor)
+
+	if cupsdSocketSourceDir == "" {
+		// no other rules, this is the legacy slot without the additional
+		// attribute
+		return nil
+	}
+
+	// add rules to access the socket dir from the slot location directly
+	// this is necessary otherwise clients get denials like this:
+	// apparmor="DENIED" operation="connect"
+	// profile="snap.test-snapd-cups-consumer.bin"
+	// name="/var/snap/test-snapd-cups-provider/common/cups.sock"
+	// pid=3195747 comm="nc" requested_mask="wr" denied_mask="wr" fsuid=0 ouid=0
+	// this denial is the same that would happen for the content interface, so
+	// we employ the same workaround from the content interface here too
+	spec.AddSnippet(fmt.Sprintf(`
+# In addition to the bind mount, add any AppArmor rules so that
+# snaps may directly access the slot implementation's files. Due
+# to a limitation in the kernel's LSM hooks for AF_UNIX, these
+# are needed for using named sockets within the exported
+# directory.
+%s/** mrwklix,`, cupsdSocketSourceDir))
+
+	// setup the snap-update-ns rules for bind mounting for the plugging snap
+	emit := spec.AddUpdateNSf
+
+	emit("  # Mount cupsd socket from cups snap to client snap\n")
+	emit("  mount options=(bind rw) %s -> /var/cups/,\n", cupsdSocketSourceDir)
+	emit("  umount /var/cups/,\n")
+
+	apparmor.GenWritableProfile(emit, cupsdSocketSourceDir, 1)
+	apparmor.GenWritableProfile(emit, "/var/cups", 1)
+
+	return nil
+}
+
+func (iface *cupsInterface) MountConnectedPlug(spec *mount.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	cupsdSocketSourceDir, err := validateCupsSocketDirAttr(slot, slot.Snap())
+	if err != nil {
+		return err
+	}
+
+	if cupsdSocketSourceDir == "" {
+		// no other rules, this is the legacy slot without the additional
+		// attribute
+		return nil
+	}
+
+	// add a bind mount of the cups-socket-dir to /var/cups of the plugging snap
+	return spec.AddMountEntry(osutil.MountEntry{
+		Name:    cupsdSocketSourceDir,
+		Dir:     "/var/cups/",
```

Comment:
this requires a mimic for /var, not sure what are the consequences of that in terms namespace updating bugs.

is there a different place we could put this that doesn't have this sort of side-effect?

Reply:
From the CUPS point of view the socket can be in any writable space on the system. A client uses what `CUPS_SERVER` points at. No requirements of being `/var/...`.
From the snapd point of view, we need the snapd experts to speak up ...

Reply:
@pedronis if we want the directory to be named "cups" with any prefix or suffix, then we cannot avoid creating a mimic entirely. If we wanted to avoid creating the mimic on /var/, we could go with /var/lib/cups instead, which would just create a mimic on /var/lib/, but that does involve /var/lib/snapd/ now being on a mimic (which I suppose is fine, snap apps from inside the namespace shouldn't need to care about anything in /var/lib/snapd), otherwise we could do something like /var/lib/misc/cups which would create a mimic over /var/lib/misc, but there is nothing inside that dir with core20 at least so the impact should be minimal. 

In terms of avoiding the namespace updating bugs, I will look into that situation, I don't remember off the top of my head if it is better or worse for this to be in a deeper directory or not, I think it depends on ordering to make sure that directory mimics higher up get processed before deeper ones which my other PR would take care of I think.

Reply:
We could also update `core{,18,20}` to include an empty `/var/cups` directory, similar to what we did for `/usr/share/fonts`. That would avoid the mimic code firing for virtually all snaps, while still supporting any weird custom bases that don't provide that directory.

Reply:
Personally I think the suggestion to add an empty /var/cups directory to base snaps would be best, I'm not sure there is a better directory that avoids creating a mimic entirely and still allows us to have `cups` in the directory name. 

I don't think it's better personally to go down the route of opening up `/run` for mimics FWIW, but we could explore it if folks really don't like `/var/cups` and think it really needs to be in `/run` somewhere.

Reply:
For me it is not important whether the socket directory is actually in `/var/...` or in `/run/...`, most important is for me that it all works just out-of-the-box, one installs a Snap which plugs `cups`, `cups` get auto-connected, and one can print right away.

So depends on file system conventions and also on convenience where it is easier to place the socket directory.

Reply:
do we have a plan now to add those empty dirs to the base snaps?

Reply:
Yes if there are no objections I will propose PR's to the base snaps to add /var/cups to the base snaps as empty dirs which will avoid unnecessary writable mimics, but IMHO that's an internal optimization that doesn't need to be done before this PR has landed